### PR TITLE
gh-113317: Add Codegen class to Argument Clinic

### PR DESCRIPTION
**Pull request by vstinner**
* Move ifndef_symbols, includes and add_include() from Clinic to Codegen. Add a 'codegen' (Codegen) attribute to Clinic.
* Remove libclinic.crenderdata module: move code to libclinic.codegen.
* BlockPrinter.print_block(): remove unused 'limited_capi' argument. Remove also 'core_includes' parameter.

* Issue: gh-113317

### CODE REVIEW

File: Tools/clinic/libclinic/codegen.py [hunk 1]
```diff
@@ -188,3 +265,35 @@ def dump(self) -> str:
...
+        self.limited_capi = limited_capi
+        self.ifndef_symbols: set[str] = set()
+        # dict: include name => Include instance
+        self.includes: dict[str, Include] = {}
+
+    def add_ifndef_symbol(self, name: str) -> bool:
+        if name in self.ifndef_symbols:
+            return False
+        self.ifndef_symbols.add(name)
+        return True
```
Comment (erlend-aasland) [inline Tools/clinic/libclinic/codegen.py]: I think this API is a little bit strange; would it be beneficial to split it in two APIs? One for adding the symbol and one for querying the list of symbols?

**Comment by erlend-aasland** [comment]
Overall, I like this refactoring.

File: Tools/clinic/libclinic/codegen.py (hunk 1 above)
Comment (vstinner) [reply Tools/clinic/libclinic/codegen.py]: I made the attribute private. There is no need for an API to **get** the `ifndef_symbols` set.

I also added get_includes() methods.

**Comment by vstinner** [comment]
@erlend-aasland: I updated the PR.

**Review by erlend-aasland — APPROVED** [review]

**Comment by vstinner** [comment]
Merged, thanks for the review.

**Pull request merged**